sklearn/neighbors/base.py

```diff
@@ -359,17 +360,48 @@ class from an array representing our data set and ask who's
                     X, self._fit_X, self.effective_metric_, n_jobs=n_jobs,
                     **self.effective_metric_params_)
 
-            neigh_ind = np.argpartition(dist, n_neighbors - 1, axis=1)
-            neigh_ind = neigh_ind[:, :n_neighbors]
-            # argpartition doesn't guarantee sorted order, so we sort again
-            neigh_ind = neigh_ind[
-                sample_range, np.argsort(dist[sample_range, neigh_ind])]
+            if issparse(dist):
+                print "Dist being printed \n"
+                print dist.toarray()
+                print dist.indices
```

Comment:
https://gist.github.com/thechargedneutron/c2f95ac46a9ab61beb0de8a91a9a533f
Running the above code (tests provided by @jnothman  ) on this branch produces the following `dist` matrix:
```
[[0.    0.    0.    0.    0.    0.    0.56968608    0.    0.44086692    0. ]
 [0.    0.    0.71768226    0.    0.    0.71600003    0.    0.    0.    0. ]
 [0.    0.71768226    0.    0.    0.    0.    0.51642623    0.    0.    0. ]
 [0.    0.    0.    0.    0.    0.31561065    0.    0.51627111    0.    0. ]
 [0.    0.    0.    0.    0.    0.    0.44107968    0.    0.    0.51790786]
 [0.    0.    0.    0.31561065    0.    0.    0.    0.38322922    0.    0.]
 [0.    0.    0.51642623    0.    0.44107968    0.    0.    0.    0.    0.]
 [0.    0.    0.    0.    0.    0.38322922    0.    0.    0.    0.40501049]
 [0.44086692    0.    0.    0.    0.88847197    0.    0.    0.    0.    0.]
 [0.    0.    0.    0.    0.    0.50529526    0.    0.40501049    0.    0.]]
```
And the corresponding indices of `dist` are:
```
[0 8 6 1 5 2 2 6 1 3 5 7 4 6 9 5 3 7 6 4 2 7 5 9 8 0 4 9 7 5]
```
And the expected result of indices are:
```
[[8 6 4]
 [5 2 9]
 [6 1 4]
 [5 7 4]
 [6 9 5]
 [3 7 9]
 [4 2 0]
 [5 9 3]
 [0 4 6]
 [7 5 4]]
```
But in row 1 for example, how are we supposed to get 4 as an index while it is passed as a sparse array with a zero at that position (visible from indices output). Am I missing out on something?
I know this issue is taking too long. Sorry for this.

Reply:
This dist matrix give you 2 neighbors, 3 if you include self as explicit zeros.
The expected result of indices corresponds to 3 neighbors, without self.

The fix is replacing in the gist `nn.kneighbors_graph(X_train.copy(), mode='distance')` by `nn.kneighbors_graph(None, mode='distance')`, which is the correct syntax to have 3 neighbors without self.

In `base.py` you have to uncomment:
```py
                if query_is_train:
                    # this is done to add self as nearest neighbor
                    neigh_ind = np.concatenate((sample_range, neigh_ind),
                                               axis=1)
                    neigh_ind = neigh_ind[:, :-1]
```

Reply:
oh my mistake, you are testing the explicit diagonal case...

Reply:
@TomDLT  Is the fix that you provided above in the gist still valid? (after you accepted there's fault in your above comment?) . I am not familiar with generating CSR Matrix from `kneighbors_graph`.